### PR TITLE
Fixes #6785, #7180 add log delete command and make log deletion delete only based on time and not idvisit.

### LINKED ISSUE #6785: Implement CLI command to precisely delete raw data by date and idsite
There are some Piwik instances that are heavily using Log Analytics/Log Replay feature. However in some cases it's required to delete previously imported data and run import again (i.e. log files were incomplete or corrupted).

It would be great improvement in process, if there could be a CLI method deleting properly raw data from all log_* tables by given date and idsite.
Such command would perform timestamp shift between site timezone and UTC and purge data properly without manual action on admin side.

### PR DESCRIPTION
**Pull request by diosmosis**
This PR fixes #6785 and #7180. It introduces a new `logs:delete` command and makes all visit deletion (including the log data purger used by PrivacyManager) visit based on required criteria instead of the max idvisit.

List of changes:
- New LogDeleter service that is used by `logs:delete` and LogDataPurger in PrivacyManager.
- New method in RawLogDao that iterates over log tables based on a set of criteria. The criteria is an array describing a list of SQL conditions. I think this should be replaced eventually by a segment string, since they do the same thing, but that would require more platform infrastructure than Piwik currently has.
- The new method in RawLogDao is used by the `usercountry:attribute` command. The looping logic was also moved to VisitorGeolocator, a late finishing touch for the command.
- Removed the testing hack from LogDataPurger. Originally, to test a concurrency issue, two events were fired in LogDataPurger. Now, we use DI.
- I added the ability to override DI config in IntegrationTestCase descendants.
- I moved log inserting test code from VisitorGeolocatorTest to a new LogHelper class (in Piwik\Tests\Framework) and re-used for testing LogDeleter.
- Moved the unused log action deletion methods to RawLogDao.

TODO:
- [x] tests for new VisitorGelocator method
- [x] tests for LogPurger service
- ~~[ ] tests for new RawLogDao methods~~
- [x] benchmark command + log purging on large instance
- [x] add test for #7180

### CODE REVIEW

**Comment by mattab** [comment]
@diosmosis Sorry for the delay in getting to review this. This Pull request is high value and fixes a major bug that can lead to data loss :+1: 

Can you rebase it (I think work done in #7964 is relevant to this PR). After rebase let me know I will review and possibly merge. thanks

**Comment by diosmosis** [comment]
@mattab The rebase was very complicated, tests may not pass (will have to check tomorrow). Please do a thorough review.

File: tests/PHPUnit/Framework/TestCase/IntegrationTestCase.php [hunk 1]
```diff
@@ -86,6 +88,15 @@ public function setUp()
             self::restoreDbTables(self::$tableData);
         }
 
+        $extraDefinitions = $this->provideContainerConfig();
```
Comment (mattab) [inline tests/PHPUnit/Framework/TestCase/IntegrationTestCase.php]: Is this change expected to be in this pull request (it seems unrelated). Either way, if the code is correct, it's fine to have it here, just double checking

File: core/DataAccess/RawLogDao.php [hunk 2]
```diff
@@ -45,38 +60,152 @@ public function updateConversions(array $values, $idVisit)
...
-    public function getVisitsWithDatesLimit($from, $to, $fields = array(), $fromId = 0, $limit = 1000)
+    public function countVisitsWithDatesLimit($from, $to)
     {
-        $sql = "SELECT " . implode(', ', $fields)
+        $sql = "SELECT COUNT(*) AS num_rows"
              . " FROM " . Common::prefixTable('log_visit')
-             . " WHERE visit_first_action_time >= ? AND visit_last_action_time < ?"
-             . " AND idvisit > ?"
-             . sprintf(" LIMIT %d", $limit);
+             . " WHERE visit_first_action_time >= ? AND visit_last_action_time < ?";
```
Comment (mattab) [inline core/DataAccess/RawLogDao.php]: here I'd be tempted to change `visit_first_action_time` to `visit_last_action_time` so this query uses the index and is faster.

Comment (mattab) [inline core/DataAccess/RawLogDao.php]: (it hasn't changed in the PR so i'm just writing here feedback so I can remember to change it, and get your thoughts on that?)

File: plugins/CoreAdminHome/Commands/DeleteLogs.php [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        parent::__construct();
+
+        $this->logDeleter = $logDeleter ?: StaticContainer::get('Piwik\LogDeleter');
+        $this->rawLogDao = $rawLogDao ?: StaticContainer::get('Piwik\DataAccess\RawLogDao');
+    }
+
+    protected function configure()
+    {
+        $this->setName('logs:delete');
+        $this->setDescription('Delete data from one of the log tables: ' . implode(', ', self::$logTables) . '.');
```
Comment (mattab) [inline plugins/CoreAdminHome/Commands/DeleteLogs.php]: suggested: `Delete data from one of the log tables` -> `Delete data from the log tables` (it does not delete only from one table but possibly from all tables)

File: plugins/CoreAdminHome/Commands/DeleteLogs.php [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        $this->addOption('dates', null, InputOption::VALUE_REQUIRED, 'Delete log data with a date within this date range. Eg, 2012-01-01,2013-01-01');
+        $this->addOption('site', null, InputOption::VALUE_REQUIRED,
```
Comment (mattab) [inline plugins/CoreAdminHome/Commands/DeleteLogs.php]: Deleting data from all websites (ie. not specifying `--site`) does not work, I get:
```
./console logs:delete --dates="2015-06-01,2015-06-07"


                              
  [InvalidArgumentException]  
  Invalid site ID:            
                              

```

File: plugins/CoreAdminHome/Commands/DeleteLogs.php [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    {
+        parent::__construct();
+
+        $this->logDeleter = $logDeleter ?: StaticContainer::get('Piwik\LogDeleter');
+        $this->rawLogDao = $rawLogDao ?: StaticContainer::get('Piwik\DataAccess\RawLogDao');
+    }
+
+    protected function configure()
+    {
+        $this->setName('logs:delete');
```
Comment (mattab) [inline plugins/CoreAdminHome/Commands/DeleteLogs.php]: fyi: i'll rename command name to `core:purge-old-logs` for consistency with `core:purge-old-archive-data`

Comment (mattab) [inline plugins/CoreAdminHome/Commands/DeleteLogs.php]: Actually would like to rename it `core:delete-logs-data` unless someone has a better idea :)

**Comment by mattab** [comment]
Looks good to me: cleaning up some tech debt, adding useful new command and fixing possibly major bug all at once. Took me long time to review it and test it but my tests worked :+1:
